enhancements/dns/externaldns-operator.md

```diff
@@ -0,0 +1,477 @@
+with the hopes of moving the operator's source code into an upstream k8s-sigs repository in the future.
+* Leverage popular frameworks and libraries , i.e. controller-runtime and kubebuilder, to simplify development
+of the operator and align with upstream communities.
+* Manage dependencies through [Go Modules](https://blog.golang.org/using-go-modules).
+* Create user and developer documentation for running ExternalDNS operator (on both OpenShift and vanilla Kubernetes).
+* Create tooling to simplify building, running, testing, etc. the operator.
+* Create tests that reduce bugs in the code and allow for continuous integration (tests should be compatible
+with OpenShift CI, in addition to vanilla kubernetes environments running in public clouds).
+* Integrate the operator with the OpenShift toolchain, i.e. [openshift/release](https://github.com/openshift/release),
+which would be required for productization.
+* Provide manifests for quickly deploying the operator (including a minimal ExternalDNS Operator CR manifest).
+* Create tooling to simplify ExternalDNS troubleshooting. For example, add support for the Operator to
+OpenShift [must-gather](https://github.com/openshift/must-gather).
+* Integrate ExternalDNS with the OpenShift monitoring toolchain. For example, add
+[metrics support](https://github.com/openshift/cluster-monitoring-operator) for the operator and its operand(s).
+
+
+In addition, the following functionality is expected as part of the operator (note that this is not an exhaustive list):
+
+* Introduce a CRD to be used by the operator to manage ExternalDNS and its dependencies.
+* Common fields in spec that define the configuration of ExternalDNS and any dependent resources,
+i.e. a secret resource containing credentials for the necessary cloud provider.
+* Common fields in spec that define how to deploy ExternalDNS.
+* Common fields in status that expose the current health & version information of ExternalDNS operands.
+* The operator will be capable of observing other resources to perform basic sequencing required for OpenShift
+integration.
+* The operator will be capable of deploying multiple instances of ExternalDNS (operands), which is necessary when
+a common domain is shared across DNS zones, or when using multiple DNS Providers. See ExternalDNS' upstream
+[AWS public/private zone tutorial docs](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/public-private-route53.md)
+as an example.
+
+Here is what a CRD instance for ExternalDNS Operator may look like:
+
+```yaml
+apiVersion: operator.externaldns.k8s.io/v1alpha1
+kind: externaldns
+metadata:
+  name: default
+  namespace: my-externaldns-namespace
+spec:
+  DNSZones:
+    - my-public-zone
+    - my-private-zone
+  domains:
+    - my-cluster.my-domain.com
+  provider:
+    type: AWS
+    credentials: my-secret-reference
+  sources:
+    - openshift-route
+    - CRD
+    - foo/service
```

Comment:
I strongly suspect that if we send this to API review, we will be asked to destructure these strings:
```yaml
    - source: openshift-route
    - source: crd
    - source: service
      namespace: foo
```
If we anticipate adding support for source-specific parameters, we might want to define a union type:
```yaml
    - route:
        ignoreHostnameAnnotation: "true"
        labelSelector:
          foo: bar
    - crd:
        resource: dnsrecords
    - service
        namespace: foo
        publishInternal: "true"
```
Are we likely to need source-specific parameters?

Reply:
Yes, I suspect a union type along the lines of what youve proposed would be necessary. I've gone ahead and updated the enhancement accordingly.